runtime/binding-kafka/src/main/java/io/aklivity/zilla/runtime/binding/kafka/internal/validator/AvroValidator.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright 2021-2023 Aklivity Inc.
+ *
+ * Aklivity licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.aklivity.zilla.runtime.binding.kafka.internal.validator;
+
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.function.LongFunction;
+import java.util.function.ToLongFunction;
+import java.util.stream.Collectors;
+
+import org.agrona.DirectBuffer;
+import org.agrona.collections.Long2ObjectHashMap;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericDatumReader;
+import org.apache.avro.io.DatumReader;
+import org.apache.avro.io.DecoderFactory;
+
+import io.aklivity.zilla.runtime.binding.kafka.internal.config.KafkaCatalogConfig;
+import io.aklivity.zilla.runtime.binding.kafka.internal.validator.config.AvroValidatorConfig;
+import io.aklivity.zilla.runtime.engine.catalog.CatalogHandler;
+
+public final class AvroValidator implements Validator
+{
+    private static final byte MAGIC_BYTE = 0x0;
+
+    private final List<KafkaCatalogConfig> catalogs;
+    private final Long2ObjectHashMap<CatalogHandler> handlersById;
+    private final DecoderFactory decoder;
+    private DatumReader reader;
+
+    public AvroValidator(
+        AvroValidatorConfig config,
+        ToLongFunction<String> resolveId,
+        LongFunction<CatalogHandler> supplyCatalog)
+    {
+        this.handlersById = new Long2ObjectHashMap<>();
+        this.decoder = DecoderFactory.get();
+        this.catalogs = config.catalogList.stream().map(c ->
+        {
+            c.catalogId = resolveId.applyAsLong(c.name);
```

Comment:
I see now why we still pass `resolveId` down the stack.

I suspect this would only go away if we can move `validators` to a first class concept in the config, managed by the engine, so it can handle this for us.